iree/compiler/Dialect/VM/Target/C/test/global_ops.mlir

```diff
@@ -0,0 +1,38 @@
+// RUN: iree-translate -iree-vm-ir-to-c-module %s | IreeFileCheck %s
+
+vm.module @global_ops {
+  // check the generated state struct
+  // CHECK-LABEL: struct global_ops_state_s {
+  // CHECK-NEXT: iree_allocator_t allocator;
+  // CHECK-NEXT: int32_t c42;
+  // CHECK-NEXT: int32_t c107_mut;
+  // CHECK-NEXT: };
```

Comment:
@benvanik I remember you suggested we could model this the same as in the interpreter, which uses a flat byte buffer I think. I just wanted to know if you have a strong opinion on which representation we should use. :)

Reply:
yep - not so much opinion as requirement, so you'll have to change it :)

Supporting indirect accesses requires that we have a reliable way of taking an arbitrary integer address to a safe memory location within the global data storage. That's not possible with a per-field modeling like this for a few reasons including padding that may be added by the compiler and aliasing through different types (loading an i32 as an f32 with bitcasting semantics). As soon as you support `vm.global.load.indirect.i32` you'll need to change to this and the addressing ops all assume the offsets are encoded as byte offsets as computed by the OrdinalAllocation pass. Add to this i64/f32/f64/etc support (where you now have differing sizes for each primitive element) and things get more complicated, as it's valid (and useful) for the compiler to be able to address the values possibly at differing offsets than the native type (like loading an i32 from the lower bits of an i64 to perform a truncating load, etc). There's no op for it yet but globals can also hold arbitrary byte buffers (`vm.rwdata`, to be added alongside `vm.rodata`) and any range of globals can be accessed as views into the global space (getting a `vm.ref<!iree.byte_buffer>` to pass to HAL methods/etc). The byte ordering here becomes critical.

Refs as well can't be individual fields for the same indirect load/store reasons. Instead of bytes though that is just an array that can be indexed as the size of iree_vm_ref_t is variable and we don't want to allow byte offsetting into them.

So you're going to need an `uint8_t rwdata[...]` and `iree_vm_ref_t refs[...];`. The `ordinal` on the global in the IR is a byte offset for the primitives and an array offset for the refs. You can get the size/counts from
https://github.com/google/iree/blob/5ddd22a5b845192498a64ba426a1000abb4debe0/iree/compiler/Dialect/VM/Target/Bytecode/BytecodeModuleTarget.cpp#L76-L96 - you can copy that and/or we could move those counts to attributes on the vm.module that gets added by https://github.com/google/iree/blob/01b89b989ab2433cbccdbd9463ba8bd67d0b5315/iree/compiler/Dialect/VM/Transforms/OrdinalAllocation.cpp#L65 (as that's the pass that computes the ordinals and where this information is known).